lib/dash/content_protection.js

```diff
@@ -262,6 +262,180 @@ shaka.dash.ContentProtection.parseFromRepresentation = function(
 };
 
 
+/**
+ * Gets a Widevine license URL from a content protection element
+ * containing a custom `ms:laurl` element
+ *
+ * @param {shaka.dash.ContentProtection.Element} element
+ * @return {string}
+ */
+shaka.dash.ContentProtection.getWidevineLicenseUrl = function(element) {
+  const mslaurlNode = shaka.util.XmlUtils.findChild(element.node, 'ms:laurl');
+  if (mslaurlNode) {
+    return mslaurlNode.getAttribute('licenseUrl') || '';
+  }
+  return '';
+};
+
+
+/**
+* @typedef {{
+*   type: number,
+*   value: !Uint16Array
+* }}
+*
+* @description
+* The parsed result of a PlayReady object record.
+*
+* @property {number} type
+*   Type of data stored in the record.
+* @property {!Uint16Array} value
+*   Record content.
+*/
+shaka.dash.ContentProtection.PlayReadyRecord;
+
+/**
+ * Enum for PlayReady record types.
+ * @enum {number}
+ */
+shaka.dash.ContentProtection.PLAYREADY_RECORD_TYPES = {
+  RIGHTS_MANAGEMENT: 0x001,
+  RESERVED: 0x002,
+  EMBEDDED_LICENSE: 0x003,
+};
+
+
+/**
+* Parses a Uint16Array for PlayReady Object Records.
+* Each PRO Record is preceeded by it's PlayReady
+* Record type and length in bytes.
+*
+* PlayReady Object Record format: https://goo.gl/FTcu46
+*
+* @param {!Uint16Array} recordData
+* @return {!Array.<shaka.dash.ContentProtection.PlayReadyRecord>}
+* @private
+*/
+shaka.dash.ContentProtection.parseMsProRecords_ = function(
+  recordData) {
+  let head = 0;
+  let records = [];
+
+  while (head < recordData.length - 1) {
+    const type = recordData[head];
+    const byteLength = recordData[head + 1];
+
+    goog.asserts.assert(
+      (byteLength & 1) === 0,
+      'expected byteLength to be an even number');
+
+    // Convert from bytes to WORD length
+    const wordLength = byteLength / 2;
+    const end = wordLength + head + 2;
+
+    // subarray end is exclusive
+    // + 2 to skip over type and byteLength
+    const recordValue = recordData.subarray(head + 2, end);
+
+    // const recordString = shaka.util.StringUtils.fromUTF8(recordValue);
+    records.push({
+      type: type,
+      value: recordValue,
+    });
+
+    head = end;
+  }
+
+  return records;
+};
+
+
+/**
+* Parses a Uint16Array for PlayReady Objects.  The data
+* should contain a 32-bit integer indicating the length of
+* the PRO in bytes.  Following that, a 16-bit integer for
+* the number of PlayReady Object Records in the PRO.  Lastly,
+* a byte array of the PRO Records themselves.
+*
+* PlayReady Object format: https://goo.gl/W8yAN4
+*
+* @param {!Uint16Array} data
```

Comment:
`Uint16Array`  can represent a sub-array, but this code doesn't work for it.  So I suggest having this accept an `ArrayBuffer` to make that requirement clear.